pkg/component/controller/cplb_unix.go

```diff
@@ -274,31 +268,25 @@ func (*Keepalived) getLinkAddresses(link netlink.Link) ([]netlink.Addr, []string
 }
 
 func (k *Keepalived) generateKeepalivedTemplate() error {
-	f, err := os.OpenFile(k.K0sVars.KeepalivedConfigFile, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, fs.FileMode(0500))
-	if err != nil {
-		return fmt.Errorf("failed to open keepalived config file: %w", err)
-	}
-	defer f.Close()
-
-	template, err := template.New("keepalived").Parse(keepalivedConfigTemplate)
-	if err != nil {
-		return fmt.Errorf("failed to parse keepalived template: %w", err)
-	}
-
+	template := template.Must(template.New("keepalived").Parse(keepalivedConfigTemplate))
 	kc := keepalivedConfig{
 		VRRPInstances: k.Config.VRRPInstances,
 	}
-	if err = template.Execute(f, kc); err != nil {
-		return fmt.Errorf("failed to execute keepalived template: %w", err)
-	}
 
-	// TODO: Do we really need to this every single time?
-	if err = os.Chown(k.K0sVars.KeepalivedConfigFile, k.uid, -1); err != nil {
-		return fmt.Errorf("failed to chown keepalived config file: %w", err)
-	}
-	if err = os.Chmod(k.K0sVars.KeepalivedConfigFile, fs.FileMode(0400)); err != nil {
-		return fmt.Errorf("failed to chmod keepalived config file: %w", err)
+	if err := file.WriteAtomically(k.configFilePath, 0400, func(file io.Writer) error {
+		if err := file.(*os.File).Chown(k.uid, -1); err != nil {
```

Comment:
maybe `file.WriteAtomically`s callback should be `func (file *os.File)` if it's always a file.

Reply:
This is actually me cheating here because I know it's a file 😈

When I created the WriteAtomically function, I explicitly wanted to hide the fact that this is a file, so that nobody would do unintended things like calling file.Name(), file.Remove() and so on. The function's API has the permissions, but it's lacking ownership information. I missed that, since `os.WriteFile` is similar, probably because you cannot convey ownership information through the `open` syscall and have to do it separately? Have to think about it a bit...

Reply:
That said, I'm not even 100% sure right now that os.Rename will keep ownership info. I assume, but would I bet? For Linux and Windows?

Reply:
Reminder to self: os.Chwon/file.Chown is not a thing on Windows anyways.

Reply:
Follow-up in #4331.